### PR TITLE
feat(event-reporter): multi-sourced apps - resource git details reporting

### PR DESCRIPTION
**Pull request by oleksandr-codefresh**
Checklist:

* [ ] Either (a) I've created an [enhancement proposal](https://github.com/argoproj/argo-cd/issues/new/choose) and discussed it with the community, (b) this is a bug fix, or (c) this does not need to be in the release notes.
* [ ] The title of the PR states what changed and the related issues number (used for the release note).
* [ ] The title of the PR conforms to the [Toolchain Guide](https://argo-cd.readthedocs.io/en/latest/developer-guide/toolchain-guide/#title-of-the-pr)
* [ ] I've included "Closes [ISSUE #]" or "Fixes [ISSUE #]" in the description to automatically close the associated issue.
* [ ] I've updated both the CLI and UI to expose my feature, or I plan to submit a second PR with them.
* [ ] Does this PR require documentation updates?
* [ ] I've updated documentation as required by this PR.
* [ ] I have signed off all my commits as required by [DCO](https://github.com/argoproj/argoproj/blob/master/community/CONTRIBUTING.md#legal)
* [ ] I have written unit and/or e2e tests for my change. PRs without these are unlikely to be merged.
* [ ] My build is green ([troubleshooting builds](https://argo-cd.readthedocs.io/en/latest/developer-guide/ci/)).
* [ ] My new feature complies with the [feature status](https://github.com/argoproj/argoproj/blob/master/community/feature-status.md) guidelines.
* [ ] I have added a brief description of why this PR is necessary and/or what this PR solves.
* [ ] Optional. My organization is added to USERS.md.
* [ ] Optional. For bug fixes, I've indicated what older releases this fix should be cherry-picked into (this may or may not happen depending on risk/complexity).

### CODE REVIEW

File: event_reporter/application/client.go [hunk 1]
```diff
@@ -126,6 +126,16 @@ func (c *httpApplicationClient) GetManifests(ctx context.Context, in *appclient.
 	if in.Revision != nil {
 		params = fmt.Sprintf("%s&revision=%s", params, *in.Revision)
 	}
+	if in.SourcePositions != nil && len(in.SourcePositions) > 0 {
+		for _, sourcePosition := range in.SourcePositions {
+			params = fmt.Sprintf("%s&sourcePositions=%s", params, sourcePosition)
```
Comment (pasha-codefresh) [inline event_reporter/application/client.go]: you have linter issues here

File: event_reporter/controller/controller.go [hunk 2]
```diff
@@ -2,6 +2,7 @@ package controller
 
 import (
 	"context"
+	"github.com/argoproj/argo-cd/v2/util/db"
```
Comment (pasha-codefresh) [inline event_reporter/controller/controller.go]: this import should not be here

File: event_reporter/reporter/application_event_reporter.go [hunk 3]
```diff
@@ -9,8 +9,12 @@ import (
 	"strings"
 	"time"
 
+	"github.com/argoproj/argo-cd/v2/util/db"
+
 	"github.com/argoproj/argo-cd/v2/event_reporter/utils"
 
+	argoutils "github.com/argoproj/argo-cd/v2/util/argo"
```
Comment (pasha-codefresh) [inline event_reporter/reporter/application_event_reporter.go:16]: please fix imports

File: event_reporter/reporter/application_event_reporter_test.go [hunk 4]
```diff
@@ -4,6 +4,9 @@ import (
 	"context"
 	"encoding/json"
 	"fmt"
+	"github.com/argoproj/argo-cd/v2/util/db"
+	"github.com/argoproj/argo-cd/v2/util/settings"
```
Comment (pasha-codefresh) [inline event_reporter/reporter/application_event_reporter_test.go]: imports

File: event_reporter/reporter/event_payload.go [hunk 5]
```diff
@@ -154,6 +155,43 @@ func getResourceEventPayload(
 	return &events.Event{Payload: payloadBytes}, nil
 }
 
+func getResourceSourceRepoUrl(
+	rr *ReportedResource,
+	reportedEntityParentApp *ReportedEntityParentApp,
+) string {
+	specCopy := reportedEntityParentApp.app.Spec.DeepCopy()
```
Comment (pasha-codefresh) [inline event_reporter/reporter/event_payload.go:164]: why you need spec copy if you not modifying it?

File: event_reporter/reporter/event_payload.go [hunk 6]
```diff
@@ -154,6 +155,43 @@ func getResourceEventPayload(
...
+
+	specCopy.Sources = reportedEntityParentApp.app.Status.Sync.ComparedTo.Sources
+	specCopy.Source = reportedEntityParentApp.app.Status.Sync.ComparedTo.Source.DeepCopy()
+
+	if specCopy.HasMultipleSources() {
+		if rr.appSourceIdx == -1 {
+			return ""
+		}
+		return specCopy.GetSourcePtrByIndex(int(rr.appSourceIdx)).RepoURL
```
Comment (pasha-codefresh) [inline event_reporter/reporter/event_payload.go]: should you check index boundary also?

File: event_reporter/reporter/event_payload.go [hunk 7]
```diff
@@ -154,6 +155,43 @@ func getResourceEventPayload(
...
+	rr *ReportedResource,
+	reportedEntityParentApp *ReportedEntityParentApp,
+) string {
+	specCopy := reportedEntityParentApp.app.Spec.DeepCopy()
+
+	specCopy.Sources = reportedEntityParentApp.app.Status.Sync.ComparedTo.Sources
+	specCopy.Source = reportedEntityParentApp.app.Status.Sync.ComparedTo.Source.DeepCopy()
+
+	if specCopy.HasMultipleSources() {
+		if rr.appSourceIdx == -1 {
```
Comment (pasha-codefresh) [inline event_reporter/reporter/event_payload.go]: lets move this condition to ReportedResource as dedicated func, you use it in multiple places

File: event_reporter/utils/app_revision.go [hunk 8]
```diff
@@ -25,6 +25,13 @@ type RevisionsData struct {
 
 const annotationRevisionKey = "app.meta.revisions-metadata"
 
+func (asrm *AppSyncRevisionsMetadata) GetSyncRevisionAt(idx int) *RevisionWithMetadata {
```
Comment (pasha-codefresh) [inline event_reporter/utils/app_revision.go:28]: please cover with baseic tests

File: event_reporter/utils/app_revision.go [hunk 9]
```diff
@@ -65,14 +72,38 @@ func GetOperationRevision(a *appv1.Application) string {
...
 
-func GetOperationStateRevision(a *appv1.Application) *string {
+func GetOperationRevisions(a *appv1.Application) []string {
+	if a == nil {
+		return nil
+	}
+
+	// this value will be used in case if application hasn't resources , like gitsource
+	revisions := a.Status.Sync.Revisions
+	if a.Status.OperationState != nil && a.Status.OperationState.Operation.Sync != nil && a.Status.OperationState.Operation.Sync.Revisions != nil && len(a.Status.OperationState.Operation.Sync.Revisions) > 0 {
```
Comment (pasha-codefresh) [inline event_reporter/utils/app_revision.go]: lets change to if -> return, and tbh i would move these if condition to some function, it is just very hard to read, i would move it to our custom types that we created before

File: event_reporter/utils/app_revision.go [hunk 10]
```diff
@@ -65,14 +72,38 @@ func GetOperationRevision(a *appv1.Application) string {
 	return revision
 }
 
-func GetOperationStateRevision(a *appv1.Application) *string {
+func GetOperationRevisions(a *appv1.Application) []string {
```
Comment (pasha-codefresh) [inline event_reporter/utils/app_revision.go:83]: also cover with basic test

File: event_reporter/utils/app_revision.go [hunk 11]
```diff
@@ -65,14 +72,38 @@ func GetOperationRevision(a *appv1.Application) string {
...
+
+func GetOperationSyncResultRevision(a *appv1.Application) *string {
 	if a == nil || a.Status.OperationState == nil || a.Status.OperationState.SyncResult == nil {
 		return nil
 	}
 
 	return &a.Status.OperationState.SyncResult.Revision
 }
 
+func GetOperationSyncResultRevisions(a *appv1.Application) *[]string {
```
Comment (pasha-codefresh) [inline event_reporter/utils/app_revision.go:109]: some basic unit test

File: reposerver/repository/repository.go [hunk 12]
```diff
@@ -1812,21 +1812,18 @@ var manifestFile = regexp.MustCompile(`^.*\.(yaml|yml|json|jsonnet)$`)
 
 // findManifests looks at all yaml files in a directory and unmarshals them into a list of unstructured objects
 func findManifests(logCtx *log.Entry, appPath string, repoRoot string, env *v1alpha1.Env, directory v1alpha1.ApplicationSourceDirectory, enabledManifestGeneration map[string]bool, maxCombinedManifestQuantity resource.Quantity) ([]manifest, error) {
-	absRepoRoot, err := filepath.Abs(repoRoot)
```
Comment (pasha-codefresh) [inline reposerver/repository/repository.go]: why you changed it?

File: server/application/application.go [hunk 13]
```diff
@@ -574,6 +574,7 @@ func (s *Server) GetManifests(ctx context.Context, q *application.ApplicationMan
 				manifestInfo.Manifests[i].CompiledManifest = string(data)
 			}
 		}
+		manifests.SourcesManifestsStartingIdx = append(manifests.SourcesManifestsStartingIdx, int32(len(manifests.Manifests)))
```
Comment (pasha-codefresh) [inline server/application/application.go:577]: why you need this field? should we push it to OSS ?

**Review by pasha-codefresh — CHANGES_REQUESTED** [review]

File: reposerver/repository/repository.go (hunk 12 above)
Comment (oleksandr-codefresh) [reply reposerver/repository/repository.go]: i just moved this line below, to have our changes in one place (will simplify new upstream version merging)

File: server/application/application.go (hunk 13 above)
Comment (oleksandr-codefresh) [reply server/application/application.go:577]: I need it to be able to report correct repo url, etc. for multi-sourced apps

File: event_reporter/reporter/event_payload.go (hunk 5 above)
Comment (oleksandr-codefresh) [reply event_reporter/reporter/event_payload.go:164]: I'm modifying in next few lines

File: event_reporter/reporter/event_payload.go (hunk 6 above)
Comment (oleksandr-codefresh) [reply event_reporter/reporter/event_payload.go]: added check to return empty line if source under such index missing

File: event_reporter/reporter/event_payload.go (hunk 7 above)
Comment (oleksandr-codefresh) [reply event_reporter/reporter/event_payload.go]: created method `appSourceIdxDetected`

File: event_reporter/utils/app_revision.go (hunk 9 above)
Comment (oleksandr-codefresh) [reply event_reporter/utils/app_revision.go]: moved to separate func

File: server/application/application.go (hunk 13 above)
Comment (oleksandr-codefresh) [reply server/application/application.go:577]: I think this info is useful for OSS also

**Review by pasha-codefresh — APPROVED** [review]

**Pull request merged**